File: fast_waves_SC_tridiag.py
from stencil_kernel import *
import sys
import numpy
import math



# --------------------------------------------------------------
# DEFINITION of the FastWavesSCTridiag stencil object
#
class FastWavesSCTridiag (StencilKernel):
    """
    Class definition of the FastWavesSCTridiag stencil.-
    """
    def __init__ (self, domain):
        #
        # call the parent's constructor
        #
        super.__init__ (self)

        #
        # output fields
        #
        self.y   = IJKRealField (domain)
        self.tmp = IJKRealField (domain)
        self.bet = IJKRealField (domain)


    def _ForwardStage (self, center, in_b, in_rhs):
        """
        The 'Do' function of the Forward stage, applied over the KMinimunCenter.-
        """
        self.bet[center] = in_b[center]
        self.y[center]   = in_rhs[center] / self.bet[center]


    def _ForwardStageFull (self, center, in_a, in_b, in_c, in_rhs):
        """
        The 'Do' function of the Forward stage, applied over the FullDomain.-
        """
        self.tmp[center] = in_c.neigh (center, (0, 0, -1)) / self.bet[center]
        self.bet[center] = in_b[center] - in_a[center] * self.tmp[center]
        self.y[center]   = (in_rhs[center] - 
                            in_a[center] * self.y.neigh (center, (0, 0, -1))) / self.bet[center]


    def _BackwardStage (self, center, in_tmp):
        """
        The 'Do' function of the Backward stage.-
        """
        self.y[center] -= in_tmp.neigh (center, (0, 0, 1)) * self.y.neigh (center, (0, 0, 1))


    def kernel (self, in_a, in_b, in_c, in_rhs):
        """
        This stencil comprises two independent stages.-
        """
        for p in self.y.interior_points (sweep='cKIncrement',
                                         height='KMinimumCenter'):
            self._ForwardStage (p, in_b, in_rhs) 
                                         
        for p in self.y.interior_points (sweep='cKIncrement'):
            self._ForwardStageFull (p, in_a, in_b, in_c, in_rhs) 
        #
        # using `self.tmp' as an input field creates a dependency among the loops
        #
        for p in self.y.interior_points (sweep='cKDecrement'):
            self._BackwardStage (p, self.tmp)



# --------------------------------------------------------------
# USAGE of the FastWavesSCTridiag object defined above
#

#
# the calculation domain on which the stencil will be applied
#
calculationDomain = (8, 8, 2)

kernel = FastWavesSCTridiag (calculationDomain)

kernel.compilation.should_unroll = False
kernel.compilation.backend       = None     # runs in pure Python (default)

#
# boundaries definition in all three dimensions
#
boundaries = {'i': (0, 0),
              'j': (0, 0),
              'k': (0, 1)}
#
# data-field definitions
#
a = IJKRealField (np.random.random (0, 10, calculationDomain))
a.set_boundaries (**boundaries)

b = IJKRealField (np.random.random (0, 10, calculationDomain))
b.set_boundaries (**boundaries)

c = IJKRealField (np.random.random (0, 10, calculationDomain))
c.set_boundaries (**boundaries)

rhs = IJKRealField (np.random.random (0, 10, calculationDomain))
rhs.set_boundaries (**boundaries)

#
# boundaries for the output fields
#
kernel.y.set_boundaries   (**boundaries)
kernel.bet.set_boundaries (**boundaries)
kernel.tmp.set_boundaries (**boundaries)

#
# print out the initial state
#
print ("Initial state")
print (kernel.y)

#
# apply the stencil
#
kernel.kernel (a,
               b,
               c,
               rhs)
#
# print out the final state
#
print ("State after initial step")
print (self.y)

File: FastWavesSCTridiag.cpp
#include "StencilFramework.h"
#include "DycoreConfiguration.h"
#include "FastWavesSCTridiag.h"

enum
{
    a, b, c, rhs, tmp, bet, y
};

namespace FastWavesSCTridiagStages
{
    template <typename TEnv>
    struct Forward
    {
        STENCIL_STAGE(TEnv)

        STAGE_PARAMETER(FullDomain, y  ) // Output
        STAGE_PARAMETER(FullDomain, a  ) // Input
        STAGE_PARAMETER(FullDomain, b  ) // Input
        STAGE_PARAMETER(FullDomain, c  ) // Input
        STAGE_PARAMETER(FullDomain, rhs) // Input
        STAGE_PARAMETER(FullDomain, tmp) // Output buffer
        STAGE_PARAMETER(FullDomain, bet) // Output buffer

        __ACC__
        static void Do(Context ctx, KMinimumCenter)
        {
            ctx[bet::Center()] = ctx[b::Center()];
            ctx[y::Center()] = ctx[rhs::Center()] / ctx[bet::Center()];
        }

        __ACC__
        static void Do(Context ctx, FullDomain)
        {
            ctx[tmp::Center()] = ctx[c::At(kminus1)] / ctx[bet::Center()];
            ctx[bet::Center()] = ctx[b::Center()] - ctx[a::Center()] * ctx[tmp::Center()];
            ctx[y::Center()] = (ctx[rhs::Center()] - ctx[a::Center()] * ctx[y::At(kminus1)]) / ctx[bet::Center()];
        }
    };

    template <typename TEnv>
    struct Backward
    {
        STENCIL_STAGE(TEnv)

        STAGE_PARAMETER(FullDomain, y  ) // Output
        STAGE_PARAMETER(FullDomain, tmp) // Input buffer

        __ACC__
        static void Do(Context ctx, FullDomain)
        {
            ctx[y::Center()] -= ctx[tmp::At(kplus1)] * ctx[y::At(kplus1)];
        }
    };

} // End namespace FastWavesSCTridiagStages

FastWavesSCTridiag::FastWavesSCTridiag() {}
FastWavesSCTridiag::~FastWavesSCTridiag() {}

void FastWavesSCTridiag::Init(const DycoreConfiguration& dycoreConfiguration, DycoreRepository& dycoreRepository, TimeIntegratorRepository& timeIntegratorRepository, FastWavesSCRepository& fastWavesSCRepository)
{
    using namespace FastWavesSCTridiagStages;

    StencilCompiler::Build(
        stencil_,
        "FastWavesSCTridiag",
        dycoreRepository.calculationDomain(),
        StencilConfiguration<Real, FastWavesSCTridiagBlockSize>(),
        pack_parameters(
            /* output fields */
            Param<y, cInOut>(dycoreRepository.w_in()),
            /* input fields */
            Param<a  , cIn>(fastWavesSCRepository.lgs_a()  ),
            Param<b  , cIn>(fastWavesSCRepository.lgs_b()  ),
            Param<c  , cIn>(fastWavesSCRepository.lgs_c()  ),
            Param<rhs, cIn>(fastWavesSCRepository.lgs_rhs())
        ),
        define_temporaries(
            StencilBuffer<tmp, Real, KRange<FullDomain,0,1> >(),
            StageVariable<bet, Real>()
        ),
        define_loops(
            define_sweep<cKIncrement>(
                define_stages(
                    StencilStage<Forward, IJRange<cComplete,0,0,0,0>, KRange<FullDomain,0,1> >()
                )
            )
            ,
            define_sweep<cKDecrement>(
                define_stages(
                    StencilStage<Backward, IJRange<cComplete,0,0,0,0>, KRange<FullDomain,0,0> >()
                )
            )
        )
    );
}

void FastWavesSCTridiag::Apply()
{
    stencil_.Apply();
}

